File: src/main/java/gg/octave/bot/utils/PlaylistUtils.java
package gg.octave.bot.utils;

import com.sedmelluq.discord.lavaplayer.player.AudioPlayerManager;
import com.sedmelluq.discord.lavaplayer.tools.io.MessageInput;
import com.sedmelluq.discord.lavaplayer.tools.io.MessageOutput;
import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.BasicAudioPlaylist;
import gg.octave.bot.Launcher;
import org.json.JSONArray;
import org.json.JSONObject;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.util.*;
import java.util.concurrent.CopyOnWriteArrayList;
import java.util.stream.Collectors;

public class PlaylistUtils {
    public static BasicAudioPlaylist decodePlaylist(List<String> encodedTracks, String name) {
        List<AudioTrack> tracks = new CopyOnWriteArrayList<>();

        for (String encodedTrack : encodedTracks)
            tracks.add(toAudioTrack(encodedTrack));

        return new BasicAudioPlaylist(name, tracks, tracks.get(0), false);
    }

    public static BasicAudioPlaylist decodePlaylist(String jsonString) {
        JSONObject object = new JSONObject(jsonString);

        String name = object.getString("name");
        boolean isSearch = object.getBoolean("search");
        int selectedIndex = object.getInt("selected");

        JSONArray encodedTracks = object.getJSONArray("tracks");
        List<AudioTrack> tracks = new ArrayList<>(encodedTracks.length());

        for (Object encodedTrack : encodedTracks) {
            tracks.add(toAudioTrack((String) encodedTrack));
        }

        AudioTrack selectedTrack = selectedIndex > -1
                ? tracks.get(selectedIndex)
                : null;

        return new BasicAudioPlaylist(name, tracks, selectedTrack, isSearch);
    }

    public static String toJsonString(AudioPlaylist playlist) {
        int selectedIndex = playlist.getSelectedTrack() != null
                ? playlist.getTracks().indexOf(playlist.getSelectedTrack())
                : -1;

        JSONArray tracks = new JSONArray();
        playlist.getTracks().stream()
                .map(PlaylistUtils::toBase64String)
                .forEach(tracks::put);

        JSONObject object = new JSONObject();
        object.put("name", playlist.getName());
        object.put("tracks", tracks);
        object.put("search", playlist.isSearchResult());
        object.put("selected", selectedIndex);

        return object.toString();
    }

    public static List<String> encodePlaylist(Queue<AudioTrack> playlist) {
        return playlist.stream()
                .filter(Objects::nonNull)
                .map(PlaylistUtils::toBase64String)
                .collect(Collectors.toList());
    }

    public static List<String> encodePlaylist(BasicAudioPlaylist playlist) {
        Queue<AudioTrack> tracks = new LinkedList<>(playlist.getTracks());
        return encodePlaylist(tracks);
    }

    public static AudioTrack toAudioTrack(String encoded) {
        try {
            AudioPlayerManager playerManager = Launcher.INSTANCE.getPlayers().getPlayerManager();
            byte[] b64 = Base64.getDecoder().decode(encoded);
            ByteArrayInputStream bais = new ByteArrayInputStream(b64);
            return playerManager.decodeTrack(new MessageInput(bais)).decodedTrack;
        } catch (Exception e) {
            return null;
        }
    }

    public static String toBase64String(AudioTrack track) {
        AudioPlayerManager playerManager = Launcher.INSTANCE.getPlayers().getPlayerManager();

        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
            playerManager.encodeTrack(new MessageOutput(baos), track);
            return Base64.getEncoder().encodeToString(baos.toByteArray());
        } catch (Exception e) {
            return null;
        }
    }
}

File: src/main/kotlin/gg/octave/bot/utils/PlaylistUtils.kt
package gg.octave.bot.utils

import com.sedmelluq.discord.lavaplayer.track.AudioPlaylist
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.BasicAudioPlaylist
import gg.octave.bot.Launcher.players
import org.json.JSONArray
import org.json.JSONObject

object PlaylistUtils {
    private val playerManager = players.playerManager

    // Playlists
    fun decodePlaylist(encodedTracks: List<String>, name: String): BasicAudioPlaylist {
        val decoded = encodedTracks.mapNotNull(::decodeMaybeNullAudioTrack)
        return BasicAudioPlaylist(name, decoded, decoded[0], false)
    }

    fun decodePlaylist(jsonString: String): BasicAudioPlaylist {
        val jo = JSONObject(jsonString)

        val name = jo.getString("name")
        val isSearch = jo.getBoolean("search")
        val selectedIndex = jo.getInt("selected")

        val encodedTracks = jo.getJSONArray("tracks")
        val tracks = mutableListOf<AudioTrack>()

        for (encodedTrack in encodedTracks) {
            val decodedTrack = decodeAudioTrack(encodedTrack as String) ?: continue
            tracks.add(decodedTrack)
        }

        val selectedTrack = if (selectedIndex > -1) tracks[selectedIndex] else null
        return BasicAudioPlaylist(name, tracks, selectedTrack, isSearch)
    }

    fun toJsonString(playlist: AudioPlaylist): String {
        val selectedIndex = playlist.selectedTrack?.let(playlist.tracks::indexOf) ?: -1
        val tracks = JSONArray()

        for (track in playlist.tracks) {
            val enc = encodeAudioTrack(track) ?: continue
            tracks.put(enc)
        }

        return JSONObject().apply {
            put("name", playlist.name)
            put("tracks", tracks)
            put("search", playlist.isSearchResult)
            put("selected", selectedIndex)
        }.toString()
    }

    fun encodePlaylist(playlist: BasicAudioPlaylist) = encodePlaylist(playlist.tracks)
    fun encodePlaylist(playlist: Collection<AudioTrack>): List<String> = playlist.map(::encodeAudioTrack)

    fun decodeAudioTrack(encoded: String) = playerManager.decodeTrack(encoded)!!

    // This is used at the top of the file. Don't ask :^)
    fun decodeMaybeNullAudioTrack(encoded: String) = playerManager.decodeTrack(encoded)
    fun encodeAudioTrack(track: AudioTrack) = playerManager.encodeTrack(track)

    // Misc
    private fun <T : Throwable, R> suppress(exception: Class<out T>, block: () -> R): R? {
        return try {
            block()
        } catch (e: Throwable) {
            if (!exception.isAssignableFrom(e::class.java)) {
                throw e
            }

            null
        }
    }
}
